### ISSUE TITLE
`asyncio.gather` sometimes fails

### ISSUE THREAD

**Issue by Tinche** [python/mypy#10682]
**Bug Report**

I'm adding typing to a large codebase, and have noticed weird `asyncio.gather` behavior.

**To Reproduce**

Here's a self-contained test:

```
from asyncio import gather


async def returns_dict() -> dict:
    return {}


async def returns_int() -> int:
    return 1


async def test():
    tasks = (returns_dict(), returns_int())
    reveal_type(await gather(*tasks))  # note: Revealed type is "Any"
```

**Expected Behavior**

The revealed type should be: `note: Revealed type is "Tuple[builtins.dict*[Any, Any], builtins.int*]"`

**Actual Behavior**

The correct return type is not recognized.

If the example is slightly changed:
```
from asyncio import gather


async def returns_dict() -> dict:
    return {}


async def returns_int() -> int:
    return 1


async def test():
    reveal_type(await gather(returns_dict(), returns_int()))
```

it works properly.

**Your Environment**

- Mypy version used: 0.902
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): allow_redefinition, check_untyped_defs, namespace_packages, warn_unreachable
- Python version used: 3.9.5
- Operating system and version: macOs Big Sur

**Comment by JelleZijlstra** [comment]
This is because it doesn't know which overload to pick in https://github.com/python/typeshed/blob/master/stdlib/asyncio/tasks.pyi#L48. I'm not sure there's a good solution until we have PEP 646/TypeVarTuple.

I checked and it does infer the type of `tasks` correctly: main.py:14: note: Revealed type is 'Tuple[typing.Coroutine[Any, Any, builtins.dict[Any, Any]], typing.Coroutine[Any, Any, builtins.int]]'

**Comment by Tinche** [comment]
Thanks for the comment. It's a little strange since it seems all the information needed is present. Why doesn't it know which overload to pick?

**Comment by JelleZijlstra** [comment]
True, maybe it should be able to do it since we know the *args are a tuple of known size.
